src/circuit_builder.rs

```diff
@@ -1071,13 +1053,40 @@ impl<C: HaloCurve> CircuitBuilder<C> {
         }
     }
 
+    fn append_public_input_gates(&mut self) {
+        let num_gates = self.num_gates();
+        let num_pi_gates = ceil_div_usize(self.public_input_index, NUM_WIRES);
+        for i in 0..num_pi_gates {
+            self.add_gate_no_constants(PublicInputGate::new(num_gates + i * 2));
+            self.add_gate_no_constants(BufferGate::new(num_gates + i * 2 + 1));
+        }
+    }
+
+    // Replace all `PublicInput`-type targets by their corresponding `Wire`-type targets
+    // in the copy constraints pairs.
+    pub fn route_public_inputs(&mut self, offset: usize) {
+        self.copy_constraints.iter_mut().for_each(|(x, y)| {
+            if let Target::PublicInput(pi) = x {
+                *x = pi.routable_target(offset);
+            }
+            if let Target::PublicInput(pi) = y {
+                *y = pi.routable_target(offset);
+            }
+        });
+    }
+
     pub fn build(mut self) -> Circuit<C> {
         // Since we will open each polynomial at three points outside of H, we need three random
         // values to ensure nothing is learned from the out-of-H openings.
         for _i in 0..3 {
             self.add_blinding_gate();
         }
 
+        // Append the public inputs at the end of the circuit and update the copy constraints.
+        let num_gates_without_pis = self.num_gates();
+        self.append_public_input_gates();
+        self.route_public_inputs(num_gates_without_pis);
+
```

Comment:
I've chosen to put the public input gates before the padding gates. It looks better to me, but has the disadvantage that we need to add `num_gates_without_pis` to the `Circuit` and `VerificationKey` structs to recover the position of the public input gates.
I could easily change it the have the public input gates after the padding gates, in which case `num_gates_without_pis` wouldn't be needed.